# =============================================================================
# plot_example.R
#
# Example script demonstrating how the output files of the connectivity
# enrichment analysis can be loaded and visualized.
# (1) Load AUCs and compute p-values / scores (cf. S.Figs. 27,28 of the paper)
# (2) Plot enrichment curves (cf. S.Fig. 28 of the paper)
#
# AUC : area under the curve  
#
# The script is provided "as is" under the MIT license included with the magnum
# distribution.
#
# --
# Daniel Marbach
# April 27, 2015
# =============================================================================


# -----------------------------------------------------------------------------
# Initialization

# Clean up
#rm(list=ls())
#dev.off()

# Load functions
source("plot_functions.R")

# Here we just load results from one GWAS and network kernel, but the provided
# functions also work with vectors specifying several GWASs and/or kernels.
# The following result files are loaded, which can be generated by following
# the step-by-step tutorial in the magnum user guide:
# - macular_degeneration_neovascular--smooth_muscle_cells_-_umbilical_vein_4stepKernel_alpha2.0_weighted.AUC.txt.gz
# - macular_degeneration_neovascular--smooth_muscle_cells_-_umbilical_vein_4stepKernel_alpha2.0_weighted.txt.gz

network_list <- c('01_neurons_fetal_brain', 
  '02_nervous_system_adult_hindbrain',
  '03_adult_forebrain',
  '04_mesenchymal_mixed',
  '05_sarcoma',
  '06_endothelial_cells',
  '07_mesenchymal_stem_smooth_muscle_cells',
  '08_connective_tissue_muscle_cells',
  '09_connective_tissue_integumental_cells',
  '10_lymphocytes',
  '11_myeloid_leukocytes',
  '12_lymphocytes_of_b_lineage',
  '13_lymphoma',
  '14_immune_organs',
  '15_myeloid_leukemia',
  '16_endo-epithelial_cells',
  '17_adenocarcinoma',
  '18_male_reproductive_organs',
  '19_liver_kidney',
  '20_gastrointestinal_system',
  '21_heart',
  '22_mouth_throat_skeletal_muscle_tissue',
  '23_lung',
  '24_glands_internal_genitalia',
  '25_pineal_gland_eye',
  '26_neuron-associated_cells_cancer',
  '27_astrocytes_pigment_cells',
  '28_neuroectodermal_tumors_sarcoma',
  '29_epithelial_cells',
  '30_extraembryonic_membrane',
  '31_epithelial_cells_of_kidney_uterus',
  '32_lung_epithelium_lung_cancer')

# The name of the GWASs
gwas <- "phs000007.pha001788.sum.genescores"
# The name of the kernels
kernels <- network_list
# The directory where result files are located (filenames are expected to be: 
# <gwas>--<kernel>.txt.gz, which is how magnum names the result files). Here we assume
# that the result files are located in the parent directory of the R working directory.
dir <- "magnum_output/pha1788"


# -----------------------------------------------------------------------------
# (1) Load AUCs and compute p-values / scores

# Load AUCs
aucs <- loadAllAUCs(dir=dir, gwas=gwas, kernels=kernels, numPermut=10000, k=2, logScale=TRUE)
# Compute empirical p-values
pvals <- computeEnrichmentPvals(aucs)
# Benjamini-Hochberg correction
pvals <- fdrCorrection(pvals)
# Compute scores
scores <- -log10(pvals)

# Print results
cat("p-value = ", pvals, "\n")
cat("score   = ", scores, "\n\n")


# -----------------------------------------------------------------------------
# (2) Plot connectivity enrichment curves

# Fraction of the enrichment curve included in the data files (value of the magnum
# option "--curve"). Used to correctly label x-axis and compute number of genes
# passing the genome-wide signifcance threshold in the GWAS. 
curveCutoff <- 0.2

# x-range for the enrichment curve plot (set to NULL to plot the complete curve)
xrange <- c(0,1000)
# y-range for the enrichment curve plot
yrange <- c(-0.2,0.2)
# Name of the output file
outfile <- paste('enrichment_plots/', gwas, '_', "enrichmentCurve.pdf", sep = '')

# Plot curves and compute significant number of enriched genes
sg <- generateEnrichmentCurvePlots(dir=dir, gwas=gwas, kernels=kernels,
                                   xrange=xrange, yrange=yrange,
                                   curveCutoff=curveCutoff,
                                   pvals=pvals,
                                   outfile=outfile)

